build: upgrade to go 1.15.4

otan opened the pull request:
* [x] Adjust the Pebble tests to run in new version.
* [x] Adjust version in Docker image ([source](./builder/Dockerfile#L199-L200)).
* [x] Rebuild and push the Docker image (following [Basic Process](#basic-process))
* [x] Bump the version in `builder.sh` accordingly ([source](./builder.sh#L6)).
* [x] Bump the version in `go-version-check.sh` ([source](./go-version-check.sh)), unless bumping to a new patch release.
* [x] Bump the go version in `go.mod`. You may also need to rerun `make vendor_rebuild` if vendoring has changed.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh#L40-42)).
* [x] Replace other mentions of the older version of go (grep for `golang:<old_version>` and `go<old_version>`).
* [ ] Update the `builder.dockerImage` parameter in the TeamCity [`Cockroach`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Cockroach&tab=projectParams) and [`Internal`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Internal&tab=projectParams) projects.

----

The most interesting change here was the OSX SDK needed to be bumped.
This involved some extra changes (clang-3.9, installing libtapi) and
compiling our own MacOSX SDK (we cannot re-use the Docker one anymore as
it is out of date). This SDK is currently public on our S3 bucket, but
can be made employees only if desired...

Other small changes:
* Use C.DBSlice as a cast instead, to avoid lint issue.
* Reformat string(i representing rune) to string(rune(i)) or
  strconv.Atoi.
* Fixed duration parsing error messages since Go 1.15 changed them.

Release note (general change): Upgraded CockroachDB's version of Go to
v1.15.4.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56363)
<!-- Reviewable:end -->

otan commented:
Looks like Go bumped their minimum OSX SDK version to 10.12. Our builder image is 10.10.
I cannot find a SDK with g++ enabled, and have yet to figure that out.
https://github.com/docker/golang-cross/ hasn't been updated either, and that's what we've traditionally relied on in the past.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks for doing this!

> Reviewed 21 of 21 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56363#-:-MLhcXBLCl1dEwRavkHQ:bw3kg2j)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder and @otan)

___
*[pkg/release/release.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56363#-MLhcERM7QrPNpxF3wA4:-MLhcERN7mGxTmuXM0b9:bf228x0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d58f14282a8b65168af8f3f6f16aff8da025ae4/pkg/release/release.go#L67)):*
> ```Go
> var SupportedTargets = []SupportedTarget{
> 	{BuildType: "linux-gnu", Suffix: ".linux-2.6.32-gnu-amd64"},
> 	// TODO(#release): The architecture is at least 10.10 until v20.2 and 10.15 v21.1 and after.
> ```

nit: should be something like "10.15 for v21.1 and after"


<!-- Sent from Reviewable.io -->

otan commented:
thanks! i'm going to wait for @jlinder to sign off too as the Dockerfile changes are non-trivial and wanted to make sure the mirror we chose is what we want

otan commented:
ping @jlinder if you get a chance :) !

jlinder commented on build/builder/Dockerfile:
```diff
@@ -115,13 +115,20 @@ RUN mkdir ncurses \
-    python
+    python \
+    libxml2-dev \
+    libssl-dev \
```
alphabetize the packages?

jlinder reviewed (approved):
One thing I'd advocate for changing. Otherwise LGTM.

otan commented on pkg/release/release.go:
```diff
@@ -64,6 +64,9 @@ type SupportedTarget struct {
 	{BuildType: "linux-gnu", Suffix: ".linux-2.6.32-gnu-amd64"},
+	// TODO(#release): The architecture is at least 10.10 until v20.2 and 10.15 for v21.1 and after.
+	// However, this seems to be hardcoded all over the place (in particular, roachprod stage),
+	// so keeping the 10.9 standard for now.
```
changing this would make `roachprod stage` code more complex as it requires switching versions by code, e.g. 19.2 => 10.9, 21.1 => 10.15, etc... (and this is supposed to be by SHA).

not on for this work atm, so leaving it as a TODO.

otan committed:
build: upgrade to go 1.15.4

The most interesting change here was the OSX SDK needed to be bumped.
This involved some extra changes (clang-3.9, installing libtapi) and
compiling our own MacOSX SDK (we cannot re-use the Docker one anymore as
it is out of date). This SDK is currently public on our S3 bucket, but
can be made employees only if desired...

Other small changes:
* Use C.DBSlice as a cast instead, to avoid lint issue.
* Reformat string(i representing rune) to string(rune(i)) or
  strconv.Atoi.
* Fixed duration parsing error messages since Go 1.15 changed them.

Release note (general change): Upgraded CockroachDB's version of Go to
v1.15.4.

otan commented:
bors r=jlinder,nvanbenschoten
